packages/react/src/components/NumberInput/NumberInput.js

```diff
@@ -254,7 +265,10 @@ class NumberInput extends Component {
       min,
       step,
       onChange: this.handleChange,
-      value: this.state.value,
+      value:
+        useControlledStateWithEventListener && this.props.onChange
```

Comment:
I don't think the presence of an `onChange` prop should be the switch that turns on controlled mode. There's many cases where developers would want to be informed 'onChange' but not control the actual value of the input themselves. Things like clearing form validation notifications, displaying computed values from inputs etc.